Step through `total_steps` instead of epochs.

ikanher opened the pull request:
## Types of changes

<!--- What types of changes does your code introduce? Please mark an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs change / refactoring / dependency upgrade

## Motivation and Context / Related issue

<!--- What problem does it solve? -->

Currently only sample rates of 1/num_batches are supported due to the way the data loaders are handled.

To have total control over sample rate, added a parameter `total_steps` to make_private and make_private_with_epsilon. When the parameter is set we will step `total_steps` of times through the data loader instead of approximated epochs. The sample rate is also exactly set to q = B/N.

<!--- Please link to an existing issue here if one exists. -->
<!--- (we recommend to have an existing issue for each pull request) -->

## How Has This Been Tested (if it applies)

<!--- Please describe here how your modifications have been tested. -->

Added a new test `test_make_private_with_epsilon_with_total_steps` to tests/privacy_engine_test.py.

It tests a few different `total_steps`, `data_size`, and `batch_size` combinations and checks that the accounting history in fact contains the correct sample rates. In addition it also check the `target_epsilon` according to the existing `test_make_private_with_epsilon` test.
 
## Checklist

<!--- Please go over all the following points, and mark an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] The documentation is up-to-date with the changes I made.
- [ ] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [] All tests passed, and additional code has been covered with new tests.

We have been failing to run the Opacus test suite without errors.

However, the change did not introduce any _new_ failing tests.

Solosneros committed:
fix: determine points for prv mean computation based on inputs

ikanher committed:
Step through `total_steps` instead of epochs.

- Added a parameter `total_steps` to make_private and
  make_private_with_epsilon. This enables the stepping through the data
  loader for `total_steps` number of times instead of stepping through
  approximately `epochs` epochs.
- More importantly, this enables us to use sample rate q = B/N. Whereas
  in the earlier implementation only sample rates of 1/j where j is a
  positive integer were supported.

ikanher committed:
Check that epochs is one when total_steps is set.

- Added a check to `make_private_with_epsilon` that epochs must be
  exactly 1 when total_steps are set.

Solosneros commented on opacus/tests/privacy_engine_test.py:
```diff
@@ -575,6 +577,60 @@ def test_make_private_with_epsilon(self):
+        model, optimizer, dl = self._init_vanilla_training()
+        target_eps = 2.0
+        target_delta = 1e-5
+        epochs = 1
```
What is this epochs=1 doing here?

ikanher replied:
Well, epochs have to be set to 1 if total_steps are set for the accounting to work, because we are using sample rate q = B/N?

In fact, I realized yesterday someone could accidentally set epochs to something else and added this to `make_private_with_epsilon`:

```
            if epochs != 1:
                raise ValueError("Epochs must be 1 when 'total_steps' is set")
```

What's your take on this? Force set epochs to 1 when `total_steps` is set?

ikanher replied:
Based on our discussion changed code to expect that epochs is set to None when `total_steps` is set, which seems like a much better solution. Thanks!

Solosneros replied:
great!

Solosneros commented on opacus/tests/privacy_engine_test.py:
```diff
@@ -575,6 +577,60 @@ def test_make_private_with_epsilon(self):
+        data_size=st.sampled_from([50, 51]),
+    )
+    @settings(deadline=None)
+    def test_make_private_with_epsilon_with_total_steps(self, total_steps, batch_size, data_size):
```
good test!

Solosneros replied:
are they testing the accounting history with the usual `make_private_with_epsilon`?

ikanher replied:
No, they are not. Opacus devs just check the target epsilon.

Solosneros reviewed (changes requested):
I would like to get rid of this one epochs=1 in the test if possible. I think it is just left over?

ikanher committed:
When total steps is set epochs must be None.

- Earlier implementation checked that epochs equals one if total steps
  are set. Fixed this to check that epochs are _not_ set when total
  steps are set.

ikanher committed:
Add data loader length test for total steps.

- Added a test to verify that the data loader length is indeed the same
  as `total_steps` if `total_steps` is set.

ikanher committed:
Reformat with black.

Solosneros reviewed (approved):
Looks good now!
